### ISSUE TITLE
[feat] `<rh-chip>` element

### ISSUE THREAD

**Issue by dcaryll** [RedHat-UX/red-hat-design-system#1456]
### Description

RHDS needs the have the ability to filter pieces of content using "chips". This is typically a single taxonomy way of filtering versus the multiple ways of filtering that a taxonomy box offers.

https://www.patternfly.org/components/chip/

Example in production on cloud.redhat.com/learn
![Screenshot 2024-01-22 at 1 34 27 PM](https://github.com/RedHat-UX/red-hat-design-system/assets/1249747/2be3a5c2-9c7e-426a-b5fb-334303cc87e8)

### Acceptance Criteria

- [ ] Design done
- [ ] Development done

**Comment by coreyvickery** [comment]
@dcaryll Just so you're aware, here are some concepts that are being explored and reviewed.

https://www.figma.com/proto/JbdmuSkfh0CuzqUZixb1UY?node-id=1-5&locale=en

**Comment by dcaryll** [comment]
These are great explorations! Thanks for sharing.

I think it's important to note how labels and chips can be distinguished
from one another.


*Dan Caryll*
Senior Manager, UX Design and Development
Red Hat - NC
100 E Davie Street

***@***.***
M: 304-281-4098

**Comment by coreyvickery** [comment]
Update, here is where I left off.

https://www.figma.com/proto/JbdmuSkfh0CuzqUZixb1UY/rh-chip?page-id=0%3A1&type=design&node-id=61-46&viewport=114%2C135%2C1&t=yQCXPMibdoF7ajpf-1&scaling=min-zoom&mode=design

**Comment by coreyvickery** [comment]
@markcaron @hellogreg I would like to get this done in `Charmander`.

**Comment by bennypowers** [comment]
pf-chip composes into pf-text-input for things like search. will rh-chip do the same? would that change it's semantics or usability? cc @adamjohnson

**Comment by adamjohnson** [comment]
> pf-chip composes into pf-text-input for things like search. will rh-chip do the same?

Was the `<pf-text-input>` + `<pf-chip>` composition use case discussed elsewhere with regards to accessibility?

This is a common pattern seen on the web. Unfortunately, as far as I can tell, the [APG Patterns](https://www.w3.org/WAI/ARIA/apg/patterns/) don't have guidance for an input containing multiple selected items (correct me if I am wrong). [Select2](https://github.com/select2/select2) is a classic library that does this but has had [accessibility issues](https://github.com/select2/select2/issues/3744) over the years.

IMO, we should steer users away from this UI pattern until the accessibility story becomes more solidified.

**Comment by coreyvickery** [comment]
@adamjohnson I think [pf-chip](https://staging-v6.patternfly.org/components/chip) is being deprecated in PatternFly 6. I am checking with the team about this.

[Here](https://www.figma.com/proto/JbdmuSkfh0CuzqUZixb1UY/rh-chip?page-id=0%3A1&type=design&node-id=61-46&viewport=114%2C138%2C1&t=zAqAZWi6fHyPFrCM-1&scaling=min-zoom&mode=design) is a comprehensive file of what I have been working on for `rh-chip`. I just need a few folks from other design teams to approve first before you get started. There are some explorations of smaller chips for inputs on the last page.

**Comment by coreyvickery** [comment]
@adamjohnson According to the PatternFly team, `pf-chip` is being replaced by `pf-label`.

If `rh-chip` needs to be used in inputs like search, maybe we can use `rh-tag` instead or add to `rh-tag` if needed, like a more [compact version](https://staging-v6.patternfly.org/components/label#compact-labels).

**Comment by zeroedin** [comment]
@coreyvickery Just FYI had a short conversation with @eyevana about `compact` on `rh-tag` today, maybe get her use case for that as well?

**Comment by coreyvickery** [comment]
Update 5/22: Design leads will give feedback on this component on Tuesday, May 28. Marionne will present as Corey will be out.

**Comment by coreyvickery** [comment]
@markcaron Do we need anyone else to approve of the [current design](https://www.figma.com/proto/j7Rjo2W6PLloxOSXQOpll0/Red-Hat-Design-System-component-library-(v2.1.0)?page-id=3229%3A2068&node-id=3229-2069&node-type=canvas&viewport=122%2C335%2C1&t=MnFHRbfxpUqcI0i6-1&scaling=min-zoom&content-scaling=fixed) or can we proceed with development?

**Comment by coreyvickery** [comment]
@markcaron Design is done, this is ready for development.

https://www.figma.com/proto/j7Rjo2W6PLloxOSXQOpll0/Red-Hat-Design-System-component-library-(v2.1.1)?page-id=3229%3A2068&node-id=3229-2069&p=f&viewport=122%2C335%2C1&t=LsBny8NWqup2Kxoz-1&scaling=min-zoom&content-scaling=fixed

**Comment by adamjohnson** [comment]
@coreyvickery:

Thanks for putting this design together. It's very helpful! I have a few questions for you.

I put together a little [scratchpad demo](https://codepen.io/adamjohnson/pen/ogNNpKK) to help clarify some things with Chip. There are obv lots of issues with spacing and it lacks variants and whatnot, but reference it when looking over my questions below:

  * Is Chip a radio or checkbox?
    * AKA: can you only select one chip at a time or multiple?
  * Does the "X" button deselect the chip?
  * Does clicking (not on the X) deselect the chip?
  * Does clicking the "X" remove the chip from the page?
  * What is "Overflow"?
    * I'm sorry, having trouble clarifying what the intent of that is from the design
  * Right now, there's no "View All" / uncheck all button/radio/checkbox. Is this intentional?
    * [Earlier mockups](https://github.com/user-attachments/assets/2ba6e4c7-9ac2-4ca5-82aa-f51f74862375) had a "View All" included.
  * Whether or not these function like a radio or checkbox, they will still need to be paired with a [`<fieldset>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset) + [`<legend>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/legend) so that they have an accessible name. Can you add a `<legend>` to the Figma?

**Comment by coreyvickery** [comment]
@adamjohnson Thanks for the demo and writing these questions.

I'm going to check with Krystal one last time to make sure she has what she needs from a design standpoint before we continue with development. I should have an answer in a day or two.

**Comment by coreyvickery** [comment]
@adamjohnson Let's meet about the following after I talk with Krystal.

> Whether or not these function like a radio or checkbox, they will still need to be paired with a [<fieldset>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset) + [<legend>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/legend) so that they have an accessible name. Can you add a <legend> to the Figma?

**Comment by bennypowers** [comment]
@adamjohnson @coreyvickery if we also create an `<rh-chip-group>` element, with a label slot/attr pair, then we can handle the semantics internally

**Comment by marionnegp** [comment]
Chatted with Krystal, Dustin, and Grady about chip and got some feedback about the visual design:
- Close button looks kind of like it's part of the text because of the icon weight and proximity. What about using `close-circle` icon for clarity? (PF uses the `close-circle-fill` icon for their clickable labels.)
- I think the close button should be removed when it's not yet selected.
- Do we need a disabled state? (Maybe this would be a feature we add later because I don't think we have a current use case for it.)

Re: question about it functioning like a radio button or a checkbox: It sounds like search would need a chip group that functions like a radio button. Search would also need the ability to remove individual chips from a group. Could we have a chip variant for each function?

**Comment by marionnegp** [comment]
> @adamjohnson @coreyvickery if we also create an <rh-chip-group> element, with a label slot/attr pair (see screenshot in OP i.e. the label is "Filter By:"), then we can handle the semantics internally

@adamjohnson, would the "Filter by:" count as a `<legend>`?

**Comment by coreyvickery** [comment]
@adamjohnson @marionnegp @markcaron More explorations, please review ASAP.

https://www.figma.com/proto/DFdnkHihYrmiI9RuHNRAcf/rh-chip?page-id=2001%3A16&node-id=2008-10&p=f&viewport=103%2C128%2C1&t=G7TWIVonPE7n6ihK-1&scaling=min-zoom&content-scaling=fixed

**Comment by adamjohnson** [comment]
@coreyvickery: Thanks for this document! It provides some much-needed clarifications.

### RE: Truncated Chip Text

Since the text is in the light DOM and can be styled by end users, we can make this a pattern. We can include documentation regarding text truncation while recommending a maximum character count and provide clear Do/Don’t examples. Generally, truncating text via CSS hides content, which can negatively impact users. IIRC, we had a similar discussion with `rh-breadcrumbs`.

### Updates & Implementations

- I’ll adjust the "Filter by:" `<legend>` to appear on the left side of the chips.
- I’ll implement logic to show/hide the close icon based on whether a chip is selected.
- I’ll add and style a disabled chip state.
- I’ll ensure all states and colors align with the given specs (pending Marionne’s feedback).

### Concern: Overflow Chip

One concern for this release is implementing the "Overflow" chip (e.g., "View all," "5 more," or "Show less"). While this pattern is straightforward in traditional implementations, working within the constraints of web components adds complexity. Given the short time frame before release—and considering I’ll be on PTO from March 6-11—delaying this feature to the next release may help us ship the component faster while maintaining quality. With Cubone scheduled for release on April 1st, ensuring enough time for development and review is critical.

### Concern: `radio` Chip & Accessibility

I’ve been experimenting with making a chip a `radio` input, and I’m seeing similar accessibility issues that Arathy faced with `pf-radio`. Since each `<rh-chip>` is an individual radio input inside an encapsulated shadow DOM, the browser doesn’t recognize its sibling radio inputs.

When testing `<rh-chip radio>` with VO + Safari, Safari incorrectly announces "Radio, 1 of 1," and tabbing results in an improper focus order—failing [WCAG SC 2.4.3, Focus Order (Level A)](https://www.w3.org/WAI/WCAG21/Understanding/focus-order.html). Given these challenges, I recommend postponing the `radio` functionality for the next release to ensure we get the accessibility implementation right and avoid creating barriers for users.

**Comment by coreyvickery** [comment]
@adamjohnson

### Truncation
I'm fine with not truncating text.

### Overflow
I'm fine with delaying this.

### `radio`
Are you proposing no `radio` functionality, but `checkbox` instead? Or what do we do here?

Let me know if you need exact styling specs.

@marionnegp What's your pick of styling out of the four options? I personally like `v1b`.

**Comment by marionnegp** [comment]
@coreyvickery, yeah, I think `v1b` shows the clearest relationship between the unselected and selected chips.

@adamjohnson, I think I mistranscribed some notes when I said that search would need it to function like a radio button because search should allow for multiple chips to be selected/added to a group. Sorry about that!

**Comment by adamjohnson** [comment]
> Are you proposing no radio functionality, but checkbox instead? Or what do we do here?

Yes, for now rh-chip will only have checkbox functionality due to the concerns outlined above.

I'll work on getting the colors synced from `v1b` and pushed up to Github. Thanks ya'll!

**Comment by coreyvickery** [comment]
Update 3/13: Docs are in progress.

**Comment by coreyvickery** [comment]
Update 3/23: Docs are done.

https://www.figma.com/proto/UnHbsHRXYNa8rfIpCRdyyH/Element---Chip?page-id=1%3A3&node-id=1-1096&p=f&viewport=-8597%2C113%2C1&t=30RrHUBi4jt1p0vx-1&scaling=scale-down&content-scaling=fixed

**Comment by bennypowers** [comment]
Closed by #2198
